### PR TITLE
Ticket2707 add ioc samsm300

### PR DESCRIPTION
**Pull request by John-Holt-Tessella**
Add SM300 drivers

See https://github.com/ISISComputingGroup/IBEX/issues/2707

### CODE REVIEW

File: motorApp/SM300Src/SM300Driver.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/*
+Motor driver support for the SM300 controller.
+*/
+
+
+// controller-specific parameters yet
+#define NUM_SM300_PARAMS 4
```
Comment (FreddieAkeroyd) [inline motorApp/SM300Src/SM300Driver.h]: The ususal asyn way is not to hard code the number of parameters, but to define the first and last parameter names and then use "last - first + 1" so new parameters can be added without needing to (or forgetting to) update the number. In  asyn 4.32 the "paramTableSize" argument is deprecated, but i expect this has not made it into asynMotorController yet

File: motorApp/SM300Src/SM300Driver.cpp [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+#include <asynOctetSyncIO.h>
+
+#include "asynMotorController.h"
+#include "asynMotorAxis.h"
+
+#include <epicsExport.h>
+#include <errlog.h>
+#include "SM300Driver.h"
+
+#define NINT(f) (int)((f)>0 ? (f)+0.5 : (f)-0.5)
```
Comment (AdrianPotter) [inline motorApp/SM300Src/SM300Driver.cpp]: This doesn't appear to be used

File: motorApp/SM300Src/SM300Driver.cpp [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+  createParam(SM300ResetString, asynParamInt32, &reset_);
+  createParam(SM300ResetAndHomeString, asynParamInt32, &reset_and_home_);
+  createParam(SM300DisconnectString, asynParamInt32, &disconnect_);  
+  createParam(SM300ErrorCodeString, asynParamInt32, &error_code_);  
+
+  startPoller(movingPollPeriod, idlePollPeriod, 2);
+}
+
+/** Send a querry string to the controller and get a return value.
+  * Querry string is prefeced with ACK STX and postfixed with EOT
```
Comment (AdrianPotter) [inline motorApp/SM300Src/SM300Driver.cpp]: Spelling: `Querry`. Please fix this and subsequent

File: motorApp/SM300Src/SM300Driver.cpp [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+	asynStatus status = this->writeController();
+
+	//set termination character is EOT without check sum *
+	// when sending send <CR> incase this mode is switched on
+	setTerminationChars("\x06", 1, "\x04\x0D", 2);
+	if (sendCommand("PEK0")) return status;
+
+	setTerminationChars("\x06", 1, "\x04", 1);
+	// set achknowlegement on
+	if (sendCommand("PEL1")) return status;
```
Comment (AdrianPotter) [inline motorApp/SM300Src/SM300Driver.cpp]: Is there a reason for not doing this as a loop? Much easier if you ever want to change what to do for each command.

```
const char * commands[] = {
    "PEL1",
    "B/ G01",
    ...
};
int n_array = (sizeof (commands) / sizeof (const char *));
for (i = 0; i < n_array; i++) {
    if (sendCommand(commands[i]) return status;
}
```

**Review by AdrianPotter — CHANGES_REQUESTED** [review]
Couple of things to think about. Nothing major. Freddie's is probably the biggest point.

**Review by AdrianPotter — APPROVED** [review]

**Pull request merged**